src/participant/prompts/namespace.ts

```diff
@@ -39,44 +47,33 @@ User: Where is the best hummus in Berlin?
 Response:
 No names found.
 `;
-
-    // eslint-disable-next-line new-cap
-    return vscode.LanguageModelChatMessage.Assistant(prompt);
   }
 
-  static getUserPrompt(prompt: string): vscode.LanguageModelChatMessage {
-    // eslint-disable-next-line new-cap
-    return vscode.LanguageModelChatMessage.User(prompt);
+  protected getUserPrompt({ request }: NamespacePromptArgs): Promise<string> {
+    return Promise.resolve(request.prompt);
   }
 
-  static buildMessages({
+  async buildMessages({
+    connectionNames,
     context,
     request,
-    connectionNames,
-  }: {
-    request: {
-      prompt: string;
-    };
-    context: vscode.ChatContext;
-    connectionNames: string[];
-  }): vscode.LanguageModelChatMessage[] {
-    let historyMessages = getHistoryMessages({ context, connectionNames });
+  }: NamespacePromptArgs): Promise<vscode.LanguageModelChatMessage[]> {
+    let historyMessages = this.getHistoryMessages({ connectionNames, context });
     // If the current user's prompt is a connection name, and the last
     // message was to connect. We want to use the last
     // message they sent before the connection name as their prompt.
-    let userPrompt = request.prompt;
     if (connectionNames.includes(request.prompt)) {
       const previousResponse = context.history[
         context.history.length - 1
       ] as vscode.ChatResponseTurn;
-      const intent = (previousResponse?.result as ChatResult).metadata.intent;
+      const intent = (previousResponse?.result as ChatResult)?.metadata.intent;
       if (intent === 'askToConnect') {
         // Go through the history in reverse order to find the last user message.
         for (let i = historyMessages.length - 1; i >= 0; i--) {
           if (
             historyMessages[i].role === vscode.LanguageModelChatMessageRole.User
           ) {
-            userPrompt = historyMessages[i].content;
+            request.prompt = historyMessages[i].content;
```

Comment:
This is going to mutate the argument to the method. Can we avoid that? I find mutation to be a slippery slope into having things changing unexpectedly.  Maybe we can create a new prompt variable in this method that we then pass to the user prompt function.